enshrine-esdt-safe/src/from_sovereign/transfer_tokens.rs

```diff
@@ -36,15 +37,82 @@ pub trait TransferTokensModule:
             sc_panic!("Operation is not registered");
         }
 
-        let minted_operation_tokens = self.mint_tokens(&operation.tokens);
+        let (wegld_amount, checked_tokens) =
+            self.check_tokens_for_wegld_fee(operation.tokens.clone());
+
+        let minted_operation_tokens = self.mint_tokens(&checked_tokens);
         let operation_tuple = OperationTuple {
             op_hash: operation_hash,
-            operation,
+            operation: operation.clone(),
         };
 
+        self.refund_wegld(&operation.data.op_sender, wegld_amount);
+
         self.distribute_payments(hash_of_hashes, operation_tuple, minted_operation_tokens);
     }
 
+    //TODO: register_token payable endpoint
+    // require x amount wegld
+    #[endpoint(registerTokens)]
+    #[payable("*")]
+    fn register_tokens(&self, tokens: MultiValueEncoded<TokenIdentifier>) {
+        let call_payment = self.call_value().single_esdt().clone();
+        let wegld_identifier = self.wegld_identifier().get();
+
+        require!(
+            call_payment.token_identifier == wegld_identifier,
+            "WEGLD is the only token accepted as register fee"
+        );
+
+        require!(
+            call_payment.amount == DEFAULT_ISSUE_COST * tokens.len() as u64,
+            "WEGLD fee amount is not met"
+        );
+
+        for token in tokens {
+            self.paid_issued_tokens().insert(token);
+        }
+    }
+
+    fn check_tokens_for_wegld_fee(
```

Comment:
This function is not really ok. 

Let's have some requirements first:
- input: `tokens`
- output: `wegld_amount` and `remainaing_tokens`

- if pos[0] is NOT WEGLD, verify all tokens don't have prefix or were already registered - meaning is transferring mainchain or registered tokens
- if pos[0] is WEGLD and len(tokens)=1 return `(0, tokens)` - meaning WEGLD was deposited
- if pos[0] is WEGLD and len(tokens)>1 and all tokens from pos[1..] do not have prefix or were already registered return `(0, tokens)` - meaning all tokens were registered or are from mainchain and WEGLD is on position 0 randomly
- if pos[0] is WEGLD and len(tokens)>1 and NOT all tokens from pos[1..] are registered return `(DEFAULT_ISSUE_COST * nr_of_unregisted, tokens[1..]) - meaning there are tokens where issue cost was not paid, sum wegld fee for all and return all tokens but WEGLD

Reply:
for the scenario where pos[0] is NOT WEGLD, should the function just return an error when there is a token that was not registered or it has a prefix ?

Reply:
You need to return from function with some kind of error (maybe add a `bool` for output) and go through fail flow (send `execute_bridge_operation_event` and `deposit_event`)